### DISCUSSION TITLE
Project always freezing after a day or so

### DISCUSSION THREAD

**Discussion by acidkewpie** [micropython/micropython#13383] [General]
Hi, I'm building a little LED clock project with a captive WiFi portal borrowed from a blog I found, with MQTT support also added. I have a BME280 sensor and a simple LDR conencted, along with a couple of neopixel outputs.

I'm running my prototype on a Lolin / Weimos ESP32 S2 mini, and it just freezes after typically 36 hours of usage. But there's no pattern to it at all as far as I can tell. It output a status to the console in some form every second (so usually just a dot for a "tick", and I see that it just stops outputting anything at all. There's no error reported in repl, and it can be at 37 seconds into any given minute, 11 seconds, it doesn't matter.... Not on the boundary of anything, not in line with any schedule being fired off at all. As such, I cant' see how it would relate to anything actually in my code..? So whilst feeling clueless I feel like I'm left doubting MicroPython itself? 

I'm running the 1.21 firmware, the portal is taken from here https://ansonvandoren.com/posts/esp8266-captive-web-portal-part-1/, the uMQTT "Robust" library https://github.com/micropython/micropython-lib but it still acts the same when I disable it. So I'm kinda stumped how to get more information about what's going on. The LED's just freeze, and nothing ever happens again, either visually or from sending the MQTT data. repl doesn't respond... just nothing until I press the ES32's reset button. So where do I go from here, as I'd really rather stay in a python world.

I don't  think there's any use showing my ugly, code, but these are the pins I'm using if it may be relevant:

```
   36 ledm = neopixel.NeoPixel(Pin(16), 12)
   37 ledh = neopixel.NeoPixel(Pin(18), 12)
   38 leds = neopixel.NeoPixel(Pin(37), 1)
   39 ldr = ADC(12, atten=ADC.ATTN_11DB)
   40 touch_pin = TouchPad(Pin(5, mode=Pin.IN))
   41 bme = bme280.BME280(i2c=I2C(0))
````

**Comment by davefes** [comment]
I struggled for about a year to keep a generic ESP32 running 24/7.  Disabling uMQTT ... extend that approach by disabling other things.
WiFi was the main source of lock-ups for me.  "Divide and conquer".

Also, throw try/except blocks everywhere and in the except block log any errors to a file on flash.  I also changed platforms to a ESP32 SPIRAM variant.

**Reply by peterhinch** [reply]
[This doc](https://github.com/peterhinch/micropython-samples/blob/master/resilient/README.md) might shed some light.

**Reply by acidkewpie** [reply]
Ah OK, that potentially makes a lot more sense. I am firing a MQTT publish every minute, but i realise it's not actually aligned to any starting time, so can obviously happen at every 37th second in a minute... I think for me then disconnecting from MQTT after each set of publish operations is likely to be the best solution if that does indeed improve things.
